src/InstrumentationEngine/CorProfilerInfoWrapper.cpp

```diff
@@ -364,21 +364,29 @@ HRESULT MicrosoftInstrumentationEngine::CCorProfilerInfoWrapper::GetILFunctionBo
 
     CComPtr<CModuleInfo> pModuleInfo;
     CComPtr<CMethodInfo> pMethodInfo;
+
+    IMAGE_COR_ILMETHOD* pMethodHeader = nullptr;
+    ULONG cbMethodSize = 0;
+
     if (SUCCEEDED(pAppDomainCollection->GetModuleInfoById(moduleId, (IModuleInfo**)&pModuleInfo)))
     {
         hr = pModuleInfo->GetMethodInfoByToken(methodToken, &pMethodInfo);
         if (SUCCEEDED(hr) && pMethodInfo->IsInstrumented())
         {
             return pMethodInfo->GetIntermediateRenderedFunctionBody(ppMethodHeader, pcbMethodSize);
         }
+        // Method info wasn't instrumented by an instrumentation method. Ask the module info for the il.
+        // This will properly manage the cache of original il bytes.
+        IfFailRet(pModuleInfo->GetMethodIl(m_pRealCorProfilerInfo, methodToken, (IMAGE_COR_ILMETHOD**)&pMethodHeader, &cbMethodSize));
+    }
+    else
+    {
+        // If this CorProfilerInfoWrapper::GetILFunctionBody() is called during OnModuleLoadFinished
+        // instead of ModuleAttachedToAssembly callback (which can happen for raw profilers), the 
+        // assembly information does not yet exist and the ModuleInfo is not yet constructed. We will
+        // not hit our cached map of ModuleId to CModuleInfo and instead call the real profiler.
+        IfFailRet(m_pRealCorProfilerInfo->GetILFunctionBody(moduleId, methodToken, (LPCBYTE*)&pMethodHeader, &cbMethodSize));
```

Comment:
Are there any other wrapped ICorProfiler methods that rely on the moduleId cache?

Reply:
SetILFunctionBody calls if (SUCCEEDED(pAppDomainCollection->GetModuleInfoById(moduleId, (IModuleInfo**)&pModuleInfo))) and returns S_OK if that fails. No other GetModuleInfo* calls occur in this class.

Reply:
All other calls using moduleId get passed to the real profiler